### PR TITLE
New feature: transparent bridging

### PR DESCRIPTION
**Pull request by JaylinYu**
work with https://github.com/nanomq/NanoNNG/pull/1022

### CODE REVIEW

File: nanomq/apps/broker.c [hunk 1]
```diff
@@ -206,68 +206,69 @@ bridge_handler(nano_work *work)
...
+				// No change if remote topic == ""
+				if (node->forwards_list[i]->remote_topic_len != 0) {
+					publish_topic = node->forwards_list[i]->remote_topic;
+				}
+				uint8_t retain;
+				uint8_t qos;
+				retain = node->forwards_list[i]->retain == NO_RETAIN
+					? work->pub_packet->fixed_header.retain
+					: node->forwards_list[i]->retain;
+				qos    = node->forwards_list[i]->retain == NO_QOS
```
Comment (wanghaEMQ) [inline nanomq/apps/broker.c:238]: retain???

File: nanomq/apps/broker.c [hunk 2]
```diff
@@ -489,6 +491,8 @@ server_cb(void *arg)
 			    (rv = unsub_ctx_handle(work)) != 0) {
 				log_error("unsub_handler [%d]", rv);
 			}
+			// proxy unsub action to bridge
+			bridge_sub_handler(work);
```
Comment (wanghaEMQ) [inline nanomq/apps/broker.c:495]: unsub?

File: nanomq/apps/broker.c (hunk 1 above)
Comment (JaylinYu) [reply nanomq/apps/broker.c:238]: This belongs to another feature. QoS overwritten

File: nanomq/apps/broker.c (hunk 2 above)
Comment (JaylinYu) [reply nanomq/apps/broker.c:495]: same func

File: nanomq/apps/broker.c (hunk 1 above)
Comment (JaylinYu) [reply nanomq/apps/broker.c:238]: But I think i got what u mean

**Comment by rty813** [comment]
When this PR will be merged? I really need this feature.

**Comment by JaylinYu** [comment]
SInce you asked, will wrap this into 0.22.10 release

**Comment by rty813** [comment]
> SInce you asked, will wrap this into 0.22.10 release

0.22.10 does not seem to have this feature merged

**Comment by JaylinYu** [comment]
> > SInce you asked, will wrap this into 0.22.10 release
> 
> 0.22.10 does not seem to have this feature merged

Try 0.23.0 pre-release

**Comment by rty813** [comment]
@JaylinYu In version 0.23.0, I added transparent = true to the configuration file. However, when the client subscribes to a topic, nanomq does not subscribe to the same topic to the bridge.

**Comment by JaylinYu** [comment]
> @JaylinYu In version 0.23.0, I added `transparent = true` to the configuration file. However, when the client subscribes to a topic, nanomq does not subscribe to the same topic to the bridge.

It works on my side. Plz share your config and log

**Comment by rty813** [comment]
@JaylinYu 我发现，如果nanomq刚启动时，立即订阅（也许是在它连到bridge之前），就会失败。如果在nanomq启动一段时间之后再订阅，就是OK的。
失败的日志：
```
2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:324 server_cb: INIT ^^^^^^^^ ctx [11] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:324 server_cb: INIT ^^^^^^^^ ctx [12] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [14] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [16] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [18] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [20] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [22] ^^^^^^^^ 

2024-11-15 07:22:22 [1] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:333 server_cb: INIT ^^^^^^^^ extra ctx [24] ^^^^^^^^ 

2024-11-15 07:22:22 [1] INFO  /home/runner/work/nanomq/nanomq/nanomq/web_server.c:512 start_rest_server: http://0.0.0.0:8081/api/v4
2024-11-15 07:22:22 [10] INFO  /home/runner/work/nanomq/nanomq/nng/src/core/dialer.c:403 dialer_connect_cb: dialer connect cb rv0 (nil)
2024-11-15 07:22:22 [10] INFO  /home/runner/work/nanomq/nanomq/nng/src/core/dialer.c:403 dialer_connect_cb: dialer connect cb rv6 (nil)
2024-11-15 07:22:22 [13] INFO  /home/runner/work/nanomq/nanomq/nng/src/core/dialer.c:403 dialer_connect_cb: dialer connect cb rv0 (nil)
NanoMQ Broker is started successfully!
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtt/broker_tcp.c:172 tcptran_pipe_init: change p_id by hashing 797899922 rv 0
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nng/src/sp/protocol/mqtt/nmq_mqtt.c:712 nano_pipe_start: client connected! addr [127.0.0.1 port [44775]

2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx1 ^^^^

2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1584 decode_pub_message: cmd: 3, retain: 0, qos: 0, dup: 0, remaining length: 193
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1632 decode_pub_message: topic: [$SYS/brokers/connected], len: [22], qos: 0
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1677 decode_pub_message: used pos: [24]
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1689 decode_pub_message: payload: [{"username":"(null)", "ts":1731655344021,"proto_name":"MQTT","keepalive":5,"return_code":"0","proto_ver":4,"client_id":"ADAS_v2_i14","clean_start":1, "IPv4":"127.0.0.1"}], len = 169
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1240 handle_pub: pipe_info size: [0]
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:642 server_cb: WAIT ^^^^ ctx1 ^^^^
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1341 free_pub_packet: free topic
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1357 free_pub_packet: free payload
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:1363 free_pub_packet: free pub_packet
2024-11-15 07:22:24 [16] DEBUG /home/runner/work/nanomq/nanomq/nanomq/pub_handler.c:89 init_pipe_content: pub_handler: init pipe_info
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx2 ^^^^

2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/sub_handler.c:68 decode_sub_msg: remainLen: [22] packetid : [1]
2024-11-15 07:22:24 [11] INFO  /home/runner/work/nanomq/nanomq/nanomq/sub_handler.c:95 decode_sub_msg: topic: [APP2SHIP/ADAS/i14] len: [17] pid [1]
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/sub_handler.c:319 sub_ctx_handle: topicLen: [17] body: [APP2SHIP/ADAS/i14]
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/nanolib/mqtt_db.c:385 dbtree_node_new: New node: [APP2SHIP]
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/nanolib/mqtt_db.c:385 dbtree_node_new: New node: [ADAS]
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/nanolib/mqtt_db.c:385 dbtree_node_new: New node: [i14]
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/sub_handler.c:408 sub_ctx_handle: end of sub ctx handle.

2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/sub_handler.c:251 encode_suback_msg: remain: [3] varint: [3 1 0 0] len: [1] packetid: [0 1] 
2024-11-15 07:22:24 [11] DEBUG /home/runner/work/nanomq/nanomq/nanomq/apps/broker.c:740 server_cb: SEND ^^^^ ctx2 ^^^^
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1186 msquic_connection_cb: msquic_connection_cb triggered! 10
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1288 msquic_connection_cb: QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1186 msquic_connection_cb: msquic_connection_cb triggered! 0
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1191 msquic_connection_cb: [conn][0xaaab0afcba00] is Connected. Resumed Session 0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1186 msquic_connection_cb: msquic_connection_cb triggered! 7
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1291 msquic_connection_cb: QUIC_CONNECTION_EVENT_STREAMS_AVAILABLE
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1186 msquic_connection_cb: msquic_connection_cb triggered! 7
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1291 msquic_connection_cb: QUIC_CONNECTION_EVENT_STREAMS_AVAILABLE
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1666 msquic_strm_open: [sid0] quic stream opening...
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1694 msquic_strm_open: [sid0][0xffff54000b60] Done...

2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1666 msquic_strm_open: [sid1] quic stream opening...
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1694 msquic_strm_open: [sid1][0xffff54002290] Done...

2024-11-15 07:22:24 [15] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:428 quic_dialer_cb: assign 0xffff54001fe0 to substreams 0
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1666 msquic_strm_open: [sid2] quic stream opening...
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1694 msquic_strm_open: [sid2][0xffff540039b0] Done...

2024-11-15 07:22:24 [15] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:428 quic_dialer_cb: assign 0xffff54003700 to substreams 1
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1666 msquic_strm_open: [sid3] quic stream opening...
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1694 msquic_strm_open: [sid3][0xffff540050d0] Done...

2024-11-15 07:22:24 [15] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:428 quic_dialer_cb: assign 0xffff54004e20 to substreams 2
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1666 msquic_strm_open: [sid4] quic stream opening...
2024-11-15 07:22:24 [15] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1694 msquic_strm_open: [sid4][0xffff540067f0] Done...

2024-11-15 07:22:24 [15] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:428 quic_dialer_cb: assign 0xffff54006540 to substreams 3
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid0] quic_strm_cb triggered! 0 conn 0xaaab0afa4870 strm 0xffff54000b60
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1442 msquic_strm_cb: [sid0][0xffff54000b60] QUIC_STREAM_EVENT_START_COMPLETE ID0 Status0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid0] start event0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid0] end
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid1] quic_strm_cb triggered! 0 conn 0xffff54001fe0 strm 0xffff54002290
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1442 msquic_strm_cb: [sid1][0xffff54002290] QUIC_STREAM_EVENT_START_COMPLETE ID4 Status0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid1] start event0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid1] end
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid2] quic_strm_cb triggered! 0 conn 0xffff54003700 strm 0xffff540039b0
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1442 msquic_strm_cb: [sid2][0xffff540039b0] QUIC_STREAM_EVENT_START_COMPLETE ID8 Status0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid2] start event0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid2] end
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid3] quic_strm_cb triggered! 0 conn 0xffff54004e20 strm 0xffff540050d0
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1442 msquic_strm_cb: [sid3][0xffff540050d0] QUIC_STREAM_EVENT_START_COMPLETE ID12 Status0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid3] start event0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid3] end
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid4] quic_strm_cb triggered! 0 conn 0xffff54006540 strm 0xffff540067f0
2024-11-15 07:22:24 [37] INFO  /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1442 msquic_strm_cb: [sid4][0xffff540067f0] QUIC_STREAM_EVENT_START_COMPLETE ID16 Status0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid4] start event0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid4] end
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/mqtt/mqtt_codec.c:776 nni_mqtt_msg_encode_fixed_header: 22 16 0 0 0
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:973 quic_stream_dowrite: [quic dowrite][sid0] start 0xffff54000b60
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:992 quic_stream_dowrite: [sid0] buf0 sz 2
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:992 quic_stream_dowrite: [sid0] buf1 sz 22
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/mqtt/transport/quic/mqtt_quic.c:1516 mqtt_quictran_pipe_start:  start recv on sub aio 0xffff5c001808
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/mqtt/transport/quic/mqtt_quic.c:1516 mqtt_quictran_pipe_start:  start recv on sub aio 0xffff5c001e40
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/mqtt/transport/quic/mqtt_quic.c:1516 mqtt_quictran_pipe_start:  start recv on sub aio 0xffff5c002478
2024-11-15 07:22:24 [18] DEBUG /home/runner/work/nanomq/nanomq/nng/src/mqtt/transport/quic/mqtt_quic.c:1516 mqtt_quictran_pipe_start:  start recv on sub aio 0xffff5c002ab0
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1319 msquic_strm_cb: [sid0] quic_strm_cb triggered! 2 conn 0xaaab0afa4870 strm 0xffff54000b60
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:1323 msquic_strm_cb: [sid0] QUIC_STREAM_EVENT_SEND_COMPLETE!
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:605 quic_stream_cb: [quic cb][sid0] start event2
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:973 quic_stream_dowrite: [quic dowrite][sid0] start 0xffff54000b60
2024-11-15 07:22:24 [37] DEBUG /home/runner/work/nanomq/nanomq/nng/src/supplemental/quic/msquic_dial.c:684 quic_stream_cb: [quic cb][sid0] end
```

**Comment by JaylinYu** [comment]
Well there is no msg cache for subscribe Now. But I can enhance it by add another layer of cache. 

现在是没有对sub消息缓存，可以下个版本再加一个

**Comment by rty813** [comment]
> Well there is no msg cache for subscribe Now. But I can enhance it by add another layer of cache.

Version 0.23.1 does not seem to have a cache function

**Comment by JaylinYu** [comment]
> > Well there is no msg cache for subscribe Now. But I can enhance it by add another layer of cache.
> 
> Version 0.23.1 does not seem to have a cache function

I just checked, TCP bridging has it, Sub/unsub & Qos msg will be cached. Only newly added QUIC SDK doesn't .

will fixed it in 0.23.2

**Pull request closed without merging**